app/jobs/importing/hud_zip/hmis_auto_migrate_job.rb

```diff
@@ -15,14 +16,42 @@ def initialize(upload_id:, data_source_id:, deidentified: false, project_whiteli
       @project_whitelist = project_whitelist
     end
 
-    def perform
-      importer = Importers::HmisAutoMigrate::UploadedZip.new(
-        data_source_id: @data_source_id,
-        upload_id: @upload_id,
-        deidentified: @deidentified,
-        allowed_projects: @project_whitelist,
+    def before(job)
+      # Re-queue this import before processing if another import is running for the same data_source
+      # This should help prevent tying up delayed job workers that are really just waiting
+      # for the previous import to complete
+      @data_source_id = job.payload_object.instance_variable_get(:@data_source_id)
+      return unless already_running_for_data_source?
+
+      Rails.logger.info("Import of Data Source: #{@data_source_id} already running...re-queuing job for #{WAIT_MINUTES} minutes from now")
+      # when this exits, it will remove the current job from the queue, so add a new one to replace it
+      new_job = job.dup
+      new_job.update(
+        locked_at: nil,
+        locked_by: nil,
+        run_at: Time.current + WAIT_MINUTES.minutes,
+        attempts: 0,
       )
-      importer.import!
+    end
+
+    def perform
+      GrdaWarehouse::DataSource.with_advisory_lock(advisory_lock_name, timeout_seconds: 0) do
+        importer = Importers::HmisAutoMigrate::UploadedZip.new(
+          data_source_id: @data_source_id,
+          upload_id: @upload_id,
+          deidentified: @deidentified,
+          allowed_projects: @project_whitelist,
+        )
+        importer.import!
+      end
```

Comment:
I think there is a data-loss race here...
```mermaid
sequenceDiagram
  Job 1-->>Lock: already_running_for_data_source?
  Job 2-->>Lock: already running_for_data_source?
  Lock-->>Job 1: false
  Lock-->>Job 2: false
  Job 1-->>+Lock: with_advisory_lock
  Note over Job 1: Do import
  Job 2-->>Lock: with_advisory_lock
  Note over Job 2: exits with out importing
  Lock-->>-Job 1: Done
```

Instead, maybe requeue if the advisory lock fails?
```suggestion
      import_completed =  GrdaWarehouse::DataSource.with_advisory_lock(advisory_lock_name, timeout_seconds: 0) do
        importer = Importers::HmisAutoMigrate::UploadedZip.new(
          data_source_id: @data_source_id,
          upload_id: @upload_id,
          deidentified: @deidentified,
          allowed_projects: @project_whitelist,
        )
        importer.import!
      end
      
      unless imported_completed
        Rails.logger.info("Import of Data Source: #{@data_source_id} already running...re-queuing job for #{WAIT_MINUTES} minutes from now")
         # when this exits, it will remove the current job from the queue, so add a new one to replace it
         new_job = job.dup
         new_job.update(
           locked_at: nil,
           locked_by: nil,
           run_at: Time.current + WAIT_MINUTES.minutes,
           attempts: 0,
         )
      end
```

Reply:
That's kinda what I was aiming for, but there isn't any easy way to get at the job from within `perform` (or not that I could find.)  `before` gets called with a job argument, so you can do things there that you can't in the other methods.  We could just make it wait (take out the `timeout_seconds: 0` which would mean if it thought it could run, but then discovered it couldn't, it'd tie up the job worker as it does now, but it would only happen if it was trying to run two jobs and they both thought no one had the lock.

Reply:
If we could get at the job from the `perform` method, we could just requeue if we were unable to obtain the lock, no before action required.